# Ошибка

[Sumor Cloud](https://sumor.cloud) Tool.<br>
[Дополнительные документы](https://sumor.cloud/error)<br>
Библиотека обработки ошибок с поддержкой многоязычности, предопределенными кодами ошибок и передачей данных.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)<br>
[![Тест](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)<br>
[![Охват](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)<br>
[![Проверить](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Установка

```bash
npm i @sumor/error --save
```

## Предварительные требования

### Версия Node.JS

Требуется версия Node.JS 16.x или выше

### Требуется ES-модуль Node.JS

Поскольку этот пакет написан как ES-модуль, пожалуйста, измените следующий код в вашем файле `package.json`:

```json
{
  "type": "module"
}
```

## Использование

### Определение ошибки

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Пользователь не найден',
    USER_EXISTED: 'Пользователь {name} существует'
  }
})

throw new MyError('USER_NOT_FOUND')
// вывод: Error: Пользователь не найден

throw new MyError('USER_EXISTED', { name: 'Алиса' })
// вывод: Error: Пользователь Алиса существует
```

### Многоязычность

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'en', // язык по умолчанию
  code: {
    USER_NOT_FOUND: 'Пользователь не найден',
    USER_EXISTED: 'Пользователь {name} существует'
  },
  i18n: {
    zh: {
      USER_NOT_FOUND: 'Пользователь не найден',
      USER_EXISTED: 'Пользователь {name} существует'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Алиса' })
error.language = 'en' // изменить язык ошибки
console.log(error)
// вывод: Error: Пользователь Алиса существует

error.language = 'zh' // изменить язык ошибки
console.log(error)
// вывод: Error: Пользователь Алиса существует
```

### Преобразование ошибки в JSON

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Пользователь не найден',
    USER_EXISTED: 'Пользователь {name} существует'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Алиса' })
console.log(error.json())
// вывод: {"код":"USER_EXISTED","сообщение":"Пользователь Алиса существует"}
```

### Подчиненная ошибка

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Ошибка проверки поля',
    FIELD_CANNOT_EMPTY: 'Поле {name} не может быть пустым',
    FIELD_TOO_LONG: 'Поле {name} слишком длинное'
  },
  i18n: {
    zh: {
      FIELD_VERIFY_FAILED: 'Ошибка проверки поля',
      FIELD_CANNOT_EMPTY: 'Поле {name} не может быть пустым',
      FIELD_TOO_LONG: 'Поле {name} слишком длинное'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'имя пользователя' }),
  new MyError('FIELD_TOO_LONG', { name: 'пароль' })
])

console.log(error.json())
/* 
вывод: 
{
  "код":"FIELD_VERIFY_FAILED",
  "сообщение":"Ошибка проверки поля",
  "ошибки":[
    {
      "код":"FIELD_CANNOT_EMPTY",
      "сообщение":"Поле имя пользователя не может быть пустым"
    },{
      "код":"FIELD_TOO_LONG",
      "сообщение":"Поле пароль слишком длинное"
    }
  ]
}
*/

error.language = 'zh'
console.log(error.json())
/*
вывод:
{
  "код":"FIELD_VERIFY_FAILED",
  "сообщение":"Ошибка проверки поля",
  "ошибки":[
    {
      "код":"FIELD_CANNOT_EMPTY",
      "сообщение":"Поле имя пользователя не может быть пустым"
    },{
      "код":"FIELD_TOO_LONG",
      "сообщение":"Поле пароль слишком длинное"
    }
  ]
}
*/
```

### Комбинирование стандартной ошибки

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Ошибка проверки поля',
    FIELD_CANNOT_EMPTY: 'Поле {name} не может быть пустым',
    FIELD_TOO_LONG: 'Поле {name} слишком длинное'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'имя пользователя' }),
  new MyError('FIELD_TOO_LONG', { name: 'пароль' }),
  new Error('Неизвестная ошибка')
])

console.log(error.json())
/*
вывод:
{
  "код":"FIELD_VERIFY_FAILED",
  "сообщение":"Ошибка проверки поля",
  "ошибки":[
    {
      "код":"FIELD_CANNOT_EMPTY",
      "сообщение":"Поле имя пользователя не может быть пустым"
    },{
      "код":"FIELD_TOO_LONG",
      "сообщение":"Поле пароль слишком длинное"
    },{
      "код":"UNKNOWN_ERROR",
      "сообщение":"Неизвестная ошибка"
    }
  ]
}
 */
```